### PR TITLE
Feature: Use threading.locks

### LINKED ISSUE #90: Use thread-safety main worker in loop
To reduce errors due to dynamic reconfiguration, we should use a thread-safety main worker in the main loop.

### LINKED ISSUE #212: bitbots_vision: Use locks instead of flags
## Is your feature request related to a problem? Please describe.
Currently, flags are used to ensure thread safety in the main loop. This method is not very clean and still could cause threading issues, as this method is not atomic.

## Describe the solution you'd like
Use `threading.Lock` instead. (See https://stackoverflow.com/questions/3310049/proper-use-of-mutexes-in-python)

### PR DESCRIPTION
**Pull request by jaagut**
## Proposed changes
Use `threading.Lock()` in

- [x] `vision.py`
- [x] `dyanmic_color_space.py`
- [x] `color.py`

## Related issues
Closes #90.
Closes #212.

## Necessary checks
- [x] Update package version
- [x] Run `catkin build`
~- [ ] Write documentation~
- [X] Create issues for future work
- [X] Test on your machine
- [ ] Test on the robot
- [X] Put the PR on our Project board

### CODE REVIEW

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py [hunk 1]
```diff
@@ -60,16 +63,59 @@ def __init__(self):
...
-        Load and update vision config.
-        Handle config changes.
+        
+        :param Config msg: new 'vision_config'-message subscriber
+        :return: None
+        """
+        if not rospy.is_shutdown():
+            with self._transfer_reconfigure_data_mutex:
+                # Set data
+                self._transfer_reconfigure_data = msg
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/dynamic_color_space.py]:
```suggestion
        with self._transfer_reconfigure_data_mutex:
            # Set data
            self._transfer_reconfigure_data = msg
```

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py [hunk 2]
```diff
@@ -142,14 +187,18 @@ def _image_callback(self, image_msg):
...
         if 1.0 < image_age.to_sec() < 1000.0:
             rospy.logwarn('Vision: Dropped incoming Image-message, because its too old! ({} sec)'.format(image_age.to_sec()),
                           logger_throttle=2, logger_name="dynamic_color_space")
             return
 
-        self._handle_image(image_msg)
+        if not rospy.is_shutdown():
+            with self._transfer_image_msg_mutex:
+                # Transfer the image to the main thread
+                self._transfer_image_msg = image_msg
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/dynamic_color_space.py]:
```suggestion
        with self._transfer_image_msg_mutex:
            # Transfer the image to the main thread
            self._transfer_image_msg = image_msg
```

File: bitbots_vision/src/bitbots_vision/dynamic_color_space.py [hunk 3]
```diff
@@ -142,14 +187,18 @@ def _image_callback(self, image_msg):
...
+        # Drops old images and cleans up the queue.
+        # Still accepts very old images, that are most likely from ROS bags.
         image_age = rospy.get_rostime() - image_msg.header.stamp
         if 1.0 < image_age.to_sec() < 1000.0:
             rospy.logwarn('Vision: Dropped incoming Image-message, because its too old! ({} sec)'.format(image_age.to_sec()),
                           logger_throttle=2, logger_name="dynamic_color_space")
             return
 
-        self._handle_image(image_msg)
+        if not rospy.is_shutdown():
```
Comment (JanGut) [inline bitbots_vision/src/bitbots_vision/dynamic_color_space.py]: skip, if locked...

**Comment by jaagut** [comment]
Have a look at #248

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 4]
```diff
@@ -417,46 +421,56 @@ def _mask_image(self, image, color_lookup_table=None):
...
         :param np.array color_lookup_table: Optional color lookup table
         :return np.array: masked image
         """
         if color_lookup_table is None:
+            # Lookup if there is another color lookup table available
+            global _transfer_color_lookup_table_data
+            if _transfer_color_lookup_table_data is not None:
+                # Copy data from shared memory
+                with self._transfer_color_lookup_table_data_mutex:
+                    data = _transfer_color_lookup_table_data
```
Comment (Flova) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py:434]: data is not necessary

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 5]
```diff
@@ -417,46 +421,56 @@ def _mask_image(self, image, color_lookup_table=None):
         """
         Returns the color mask of the image based on the dynamic color lookup table unless other is specified
         (0 for not in color range and 255 for in color range)
-
         :param np.array image: input image
         :param np.array color_lookup_table: Optional color lookup table
         :return np.array: masked image
         """
         if color_lookup_table is None:
+            # Lookup if there is another color lookup table available
```
Comment (Flova) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py:429]: This whole part is not needed

File: bitbots_vision/src/bitbots_vision/vision_modules/color.py [hunk 6]
```diff
@@ -417,46 +421,56 @@ def _mask_image(self, image, color_lookup_table=None):
...
     def color_lookup_table_callback(self, msg):
         # type: (ColorLookupTable) -> None
         """
         This callback gets called inside the vision node, after subscriber received ColorLookupTableMessage from DynamicColorLookupTable-Node.
-
         :param ColorLookupTableMessage msg: ColorLookupTableMessage
         :return: None
         """
-        self._decode_color_lookup_table(msg)
+        global _transfer_color_lookup_table_data
```
Comment (Flova) [inline bitbots_vision/src/bitbots_vision/vision_modules/color.py:451]: Add return if thread is locked

**Review by Flova — CHANGES_REQUESTED** [review]

**Review by Flova — APPROVED** [review]

**Pull request merged**